language/vm/src/file_format.rs

```diff
@@ -1689,6 +1692,10 @@ impl CompiledModule {
 
     /// Returns the number of items of a specific `IndexKind`.
     pub fn kind_count(&self, kind: IndexKind) -> usize {
+        precondition!(match kind {
```

Comment:
I am not sure about this. If I understand it, this is making a precondition for the `unreachable` in the inner `kind_count`, however it creates this "duplication" about having the precondition disconnected from where the implementation is.
In other words changes in the inner `kind_count` may not be reflected in this precondition (or are they?).
Would it make more sense to have this in the inner `kind_count`?

Reply:
This function calls CompiledModuleMut::kind_count which has an unreachable!() clause in it that implicitly creates a precondition that the caller must uphold if it does not want to panic. Since this kind_count function is public, we do not know all of its callers during analysis. MIRAI therefore comments on this and asks that you make the implicit precondition explicit.

There is no runtime cost here. The precondition macro compiles to nothing.